```
error[E0271]: type mismatch resolving `<std::vec::IntoIter<(style::rule_tree::StyleSource, style::properties::Importance)> as std::iter::Iterator>::Item == (style::rule_tree::StyleSource, style::rule_tree::CascadeLevel)`
  --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/rule_tree/bench.rs:65:15
   |
65 |     rule_tree.insert_ordered_rules(rules.into_iter())
   |               ^^^^^^^^^^^^^^^^^^^^ expected enum `style::properties::Importance`, found enum `style::rule_tree::CascadeLevel`
   |
   = note: expected type `(style::rule_tree::StyleSource, style::properties::Importance)`
              found type `(style::rule_tree::StyleSource, style::rule_tree::CascadeLevel)`

error[E0061]: this function takes 3 parameters but 1 parameter was supplied
   --> /Users/servo/buildbot/slave/mac-dev-unit/build/tests/unit/style/stylist.rs:116:29
    |
116 |     map.get_universal_rules(&mut decls);
    |                             ^^^^^^^^^^ expected 3 parameters
```

--- a/components/script/dom/cssstyledeclaration.rs
+++ b/components/script/dom/cssstyledeclaration.rs
@@ -11,12 +11,13 @@ use dom::bindings::reflector::{DomObject, Reflector, reflect_dom_object};
 use dom::bindings::str::DOMString;
 use dom::cssrule::CSSRule;
 use dom::element::Element;
-use dom::node::{Node, NodeDamage, window_from_node};
+use dom::node::{Node, window_from_node};
 use dom::window::Window;
 use parking_lot::RwLock;
 use servo_url::ServoUrl;
 use std::ascii::AsciiExt;
 use std::sync::Arc;
+use style::attr::AttrValue;
 use style::parser::ParserContextExtraData;
 use style::properties::{Importance, PropertyDeclarationBlock, PropertyId, LonghandId, ShorthandId};
 use style::properties::{parse_one_declaration, parse_style_attribute};
@@ -53,16 +54,12 @@ impl CSSStyleOwner {
         let mut changed = true;
         match *self {
             CSSStyleOwner::Element(ref el) => {
-                let mut attr = el.style_attribute().borrow_mut();
-                let (result, needs_clear) = if attr.is_some() {
+                let mut attr = el.style_attribute().borrow_mut().take();
+                let result = if attr.is_some() {
                     let lock = attr.as_ref().unwrap();
                     let mut pdb = lock.write();
                     let result = f(&mut pdb, &mut changed);
-                    if changed {
-                        el.set_style_attr(pdb.to_css_string());
-                        el.upcast::<Node>().dirty(NodeDamage::NodeStyleDamaged);
-                    }
-                    (result, pdb.declarations.is_empty())
+                    result
                 } else {
                     let mut pdb = PropertyDeclarationBlock {
                         important_count: 0,
@@ -72,18 +69,32 @@ impl CSSStyleOwner {
 
                     // Here `changed` is somewhat silly, because we know the
                     // exact conditions under it changes.
-                    if !pdb.declarations.is_empty() {
-                        el.set_style_attr(pdb.to_css_string());
-                        el.upcast::<Node>().dirty(NodeDamage::NodeStyleDamaged);
-                        *attr = Some(Arc::new(RwLock::new(pdb)));
+                    changed = !pdb.declarations.is_empty();
+                    if changed {
+                        attr = Some(Arc::new(RwLock::new(pdb)));
                     }
 
-                    (result, false)
+                    result
                 };
 
-                if needs_clear {
-                    *attr = None;
+                if changed {
+                    // Note that there's no need to remove the attribute here if
+                    // the declaration block is empty[1], and if `attr` is
+                    // `None` it means that it necessarily didn't change, so no
+                    // need to go through all the set_attribute machinery.
+                    //
+                    // [1]: https://github.com/whatwg/html/issues/2306
+                    if let Some(pdb) = attr {
+                        let serialization = pdb.read().to_css_string();
+                        el.set_attribute(&local_name!("style"),
+                                         AttrValue::Declaration(serialization,
+                                                                pdb));
+                    }
+                } else {
+                    // Remember to put it back.
+                    *el.style_attribute().borrow_mut() = attr;
                 }
+
                 result
             }
             CSSStyleOwner::CSSRule(ref rule, ref pdb) => {

--- a/components/script/dom/document.rs
+++ b/components/script/dom/document.rs
@@ -179,7 +179,7 @@ impl PendingRestyle {
     }
 }
 
-// https://dom.spec.whatwg.org/#document
+/// https://dom.spec.whatwg.org/#document
 #[dom_struct]
 pub struct Document {
     node: Node,

--- a/components/script/dom/document.rs
+++ b/components/script/dom/document.rs
@@ -129,7 +129,7 @@ use std::sync::Arc;
 use std::time::{Duration, Instant};
 use style::attr::AttrValue;
 use style::context::{QuirksMode, ReflowGoal};
-use style::restyle_hints::RestyleHint;
+use style::restyle_hints::{RestyleHint, RESTYLE_STYLE_ATTRIBUTE};
 use style::selector_parser::{RestyleDamage, Snapshot};
 use style::str::{split_html_space_chars, str_join};
 use style::stylesheets::Stylesheet;
@@ -2142,23 +2142,32 @@ impl Document {
         RefMut::map(map, |m| m.entry(JS::from_ref(el)).or_insert_with(PendingRestyle::new))
     }
 
-    pub fn ensure_snapshot(&self, el: &Element) -> RefMut<Snapshot> {
+    pub fn element_state_will_change(&self, el: &Element) {
         let mut entry = self.ensure_pending_restyle(el);
         if entry.snapshot.is_none() {
             entry.snapshot = Some(Snapshot::new(el.html_element_in_html_document()));
         }
-        RefMut::map(entry, |e| e.snapshot.as_mut().unwrap())
-    }
-
-    pub fn element_state_will_change(&self, el: &Element) {
-        let mut snapshot = self.ensure_snapshot(el);
+        let mut snapshot = entry.snapshot.as_mut().unwrap();
         if snapshot.state.is_none() {
             snapshot.state = Some(el.state());
         }
     }
 
-    pub fn element_attr_will_change(&self, el: &Element, _attr: &Attr) {
-        let mut snapshot = self.ensure_snapshot(el);
+    pub fn element_attr_will_change(&self, el: &Element, attr: &Attr) {
+        // FIXME(emilio): Kind of a shame we have to duplicate this.
+        //
+        // I'm getting rid of the whole hashtable soon anyway, since all it does
+        // right now is populate the element restyle data in layout, and we
+        // could in theory do it in the DOM I think.
+        let mut entry = self.ensure_pending_restyle(el);
+        if entry.snapshot.is_none() {
+            entry.snapshot = Some(Snapshot::new(el.html_element_in_html_document()));
+        }
+        if attr.local_name() == &local_name!("style") {
+            entry.hint |= RESTYLE_STYLE_ATTRIBUTE;
+        }
+
+        let mut snapshot = entry.snapshot.as_mut().unwrap();
         if snapshot.attrs.is_none() {
             let attrs = el.attrs()
                           .iter()

--- a/components/script/dom/element.rs
+++ b/components/script/dom/element.rs
@@ -824,31 +824,6 @@ impl Element {
         }
     }
 
-    // this sync method is called upon modification of the style_attribute property,
-    // therefore, it should not trigger subsequent mutation events
-    pub fn set_style_attr(&self, new_value: String) {
-        let mut new_style = AttrValue::String(new_value);
-
-        if let Some(style_attr) = self.attrs.borrow().iter().find(|a| a.name() == &local_name!("style")) {
-            style_attr.swap_value(&mut new_style);
-            return;
-        }
-
-        // explicitly not calling the push_new_attribute convenience method
-        // in order to avoid triggering mutation events
-        let window = window_from_node(self);
-        let attr = Attr::new(&window,
-                             local_name!("style"),
-                             new_style,
-                             local_name!("style"),
-                             ns!(),
-                             None,
-                             Some(self));
-
-         assert!(attr.GetOwnerElement().r() == Some(self));
-         self.attrs.borrow_mut().push(JS::from_ref(&attr));
-    }
-
     pub fn serialize(&self, traversal_scope: TraversalScope) -> Fallible<DOMString> {
         let mut writer = vec![];
         match serialize(&mut writer,
@@ -2121,18 +2096,43 @@ impl VirtualMethods for Element {
         match attr.local_name() {
             &local_name!("style") => {
                 // Modifying the `style` attribute might change style.
-                *self.style_attribute.borrow_mut() =
-                    mutation.new_value(attr).map(|value| {
-                        let win = window_from_node(self);
-                        Arc::new(RwLock::new(parse_style_attribute(
-                            &value,
-                            &doc.base_url(),
-                            win.css_error_reporter(),
-                            ParserContextExtraData::default())))
-                    });
-                if node.is_in_doc() {
-                    node.dirty(NodeDamage::NodeStyleDamaged);
-                }
+                *self.style_attribute.borrow_mut() = match mutation {
+                    AttributeMutation::Set(..) => {
+                        // This is the fast path we use from
+                        // CSSStyleDeclaration.
+                        //
+                        // Juggle a bit to keep the borrow checker happy
+                        // while avoiding the extra clone.
+                        let is_declaration = match *attr.value() {
+                            AttrValue::Declaration(..) => true,
+                            _ => false,
+                        };
+
+                        let block = if is_declaration {
+                            let mut value = AttrValue::String(String::new());
+                            attr.swap_value(&mut value);
+                            let (serialization, block) = match value {
+                                AttrValue::Declaration(s, b) => (s, b),
+                                _ => unreachable!(),
+                            };
+                            let mut value = AttrValue::String(serialization);
+                            attr.swap_value(&mut value);
+                            block
+                        } else {
+                            let win = window_from_node(self);
+                            Arc::new(RwLock::new(parse_style_attribute(
+                                &attr.value(),
+                                &doc.base_url(),
+                                win.css_error_reporter(),
+                                ParserContextExtraData::default())))
+                        };
+
+                        Some(block)
+                    }
+                    AttributeMutation::Removed => {
+                        None
+                    }
+                };
             },
             &local_name!("id") => {
                 *self.id_attribute.borrow_mut() =
@@ -2716,7 +2716,7 @@ impl Element {
     }
 }
 
-#[derive(Clone, Copy, PartialEq)]
+#[derive(Clone, Copy)]
 pub enum AttributeMutation<'a> {
     /// The attribute is set, keep track of old value.
     /// https://dom.spec.whatwg.org/#attribute-is-set
@@ -2728,6 +2728,13 @@ pub enum AttributeMutation<'a> {
 }
 
 impl<'a> AttributeMutation<'a> {
+    pub fn is_removal(&self) -> bool {
+        match *self {
+            AttributeMutation::Removed => true,
+            AttributeMutation::Set(..) => false,
+        }
+    }
+
     pub fn new_value<'b>(&self, attr: &'b Attr) -> Option<Ref<'b, AttrValue>> {
         match *self {
             AttributeMutation::Set(_) => Some(attr.value()),

--- a/components/script/dom/htmllinkelement.rs
+++ b/components/script/dom/htmllinkelement.rs
@@ -166,7 +166,7 @@ impl VirtualMethods for HTMLLinkElement {
 
     fn attribute_mutated(&self, attr: &Attr, mutation: AttributeMutation) {
         self.super_type().unwrap().attribute_mutated(attr, mutation);
-        if !self.upcast::<Node>().is_in_doc() || mutation == AttributeMutation::Removed {
+        if !self.upcast::<Node>().is_in_doc() || mutation.is_removal() {
             return;
         }
 

--- a/components/style/attr.rs
+++ b/components/style/attr.rs
@@ -11,9 +11,12 @@ use app_units::Au;
 use cssparser::{self, Color, RGBA};
 use euclid::num::Zero;
 use num_traits::ToPrimitive;
+use parking_lot::RwLock;
+use properties::PropertyDeclarationBlock;
 use servo_url::ServoUrl;
 use std::ascii::AsciiExt;
 use std::str::FromStr;
+use std::sync::Arc;
 use str::{HTML_SPACE_CHARACTERS, read_exponent, read_fraction};
 use str::{read_numbers, split_commas, split_html_space_chars};
 #[cfg(not(feature = "gecko"))] use str::str_join;
@@ -30,7 +33,7 @@ pub enum LengthOrPercentageOrAuto {
     Length(Au),
 }
 
-#[derive(PartialEq, Clone, Debug)]
+#[derive(Clone, Debug)]
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub enum AttrValue {
     String(String),
@@ -43,6 +46,22 @@ pub enum AttrValue {
     Color(String, Option<RGBA>),
     Dimension(String, LengthOrPercentageOrAuto),
     Url(String, Option<ServoUrl>),
+
+    /// Note that this variant is only used transitively as a fast path to set
+    /// the property declaration block relevant to the style of an element when
+    /// set from the inline declaration of that element (that is,
+    /// `element.style`).
+    ///
+    /// This can, as of this writing, only correspond to the value of the
+    /// `style` element, and is set from its relevant CSSInlineStyleDeclaration,
+    /// and then converted to a string in Element::attribute_mutated.
+    ///
+    /// Note that we don't necessarily need to do that (we could just clone the
+    /// declaration block), but that avoids keeping a refcounted
+    /// declarationblock for longer than needed.
+    Declaration(String,
+                #[cfg_attr(feature = "servo", ignore_heap_size_of = "Arc")]
+                Arc<RwLock<PropertyDeclarationBlock>>)
 }
 
 /// Shared implementation to parse an integer according to
@@ -330,6 +349,7 @@ impl ::std::ops::Deref for AttrValue {
                 AttrValue::Color(ref value, _) |
                 AttrValue::Int(ref value, _) |
                 AttrValue::Url(ref value, _) |
+                AttrValue::Declaration(ref value, _) |
                 AttrValue::Dimension(ref value, _) => &value,
             AttrValue::Atom(ref value) => &value,
         }

--- a/components/style/restyle_hints.rs
+++ b/components/style/restyle_hints.rs
@@ -30,14 +30,23 @@ bitflags! {
     /// expensive, and so we use RestyleHints to short-circuit work we know is
     /// unnecessary.
     pub flags RestyleHint: u32 {
-        #[doc = "Rerun selector matching on the element."]
+        /// Rerun selector matching on the element.
         const RESTYLE_SELF = 0x01,
-        #[doc = "Rerun selector matching on all of the element's descendants."]
-        // NB: In Gecko, we have RESTYLE_SUBTREE which is inclusive of self, but heycam isn't aware
-        // of a good reason for that.
+
+        /// Rerun selector matching on all of the element's descendants.
+        ///
+        /// NB: In Gecko, we have RESTYLE_SUBTREE which is inclusive of self,
+        /// but heycam isn't aware of a good reason for that.
         const RESTYLE_DESCENDANTS = 0x02,
-        #[doc = "Rerun selector matching on all later siblings of the element and all of their descendants."]
+
+        /// Rerun selector matching on all later siblings of the element and all
+        /// of their descendants.
         const RESTYLE_LATER_SIBLINGS = 0x08,
+
+        /// Don't re-run selector-matching on the element, only the style
+        /// attribute has changed, and this change didn't have any other
+        /// dependencies.
+        const RESTYLE_STYLE_ATTRIBUTE = 0x10,
     }
 }
 

